### PR TITLE
[Compatibility] Added ZUNION, ZUNIONSTORE commands

### PR DESCRIPTION
**Pull request by Vijay-Nirmal**
Adding the ZUNION, ZUNIONSTORE commands to garnet

- [x] Add ZUNION, ZUNIONSTORE commands
- [x] Add Integration Test cases, ACL Test and Slot Verification Test
- [x] Add documentation

### CODE REVIEW

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs [hunk 1]
```diff
@@ -886,6 +886,155 @@ public GarnetStatus SortedSetScan<TObjectContext>(byte[] key, ref ObjectInput in
...
+            var status = GET(keys[0].ToArray(), out var firstObj, ref objectContext);
+            if (status == GarnetStatus.OK)
+            {
+                if (firstObj.garnetObject is not SortedSetObject firstSortedSet)
+                {
+                    return GarnetStatus.WRONGTYPE;
+                }
+
+                // Initialize pairs with the first set
+                pairs = new Dictionary<byte[], double>(new ByteArrayComparer());
```
Comment (TalZaccai) [inline libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:995]: Similar comment to SortedSetIntersection (SortedSetOps.cs line 959)

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs [hunk 2]
```diff
@@ -886,6 +886,155 @@ public GarnetStatus SortedSetScan<TObjectContext>(byte[] key, ref ObjectInput in
...
+
+                    if (nextObj.garnetObject is not SortedSetObject nextSortedSet)
+                    {
+                        pairs = default;
+                        return GarnetStatus.WRONGTYPE;
+                    }
+
+                    foreach (var (key, score) in nextSortedSet.Dictionary)
+                    {
+                        var weightedScore = weights != null ? score * weights[i] : score;
```
Comment (TalZaccai) [inline libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:1016]: Similar comment to SortedSetIntersection (SortedSetOps.cs line 989)

File: test/Garnet.test/RespSortedSetTests.cs [hunk 3]
```diff
@@ -1086,6 +1086,150 @@ public void CheckSortedSetDifferenceStoreWithNoMatchSE()
...
+                ClassicAssert.AreEqual(expectedScores[i], result[i].Score);
+                ClassicAssert.AreEqual(expectedElements[i], result[i].Element.ToString());
+            }
+        }
+
+        [Test]
+        [TestCase("SUM", new double[] { 3, 5, 6, 7 }, new string[] { "c", "a", "d", "b" }, Description = "Tests ZUNIONSTORE with SUM aggregate")]
+        [TestCase("MIN", new double[] { 1, 2, 3, 6 }, new string[] { "a", "b", "c", "d" }, Description = "Tests ZUNIONSTORE with MIN aggregate")]
+        [TestCase("MAX", new double[] { 3, 4, 5, 6 }, new string[] { "c", "a", "b", "d" }, Description = "Tests ZUNIONSTORE with MAX aggregate")]
+        public void CanUseZUnionStoreWithAggregateOption(string aggregateType, double[] expectedScores, string[] expectedElements)
```
Comment (TalZaccai) [inline test/Garnet.test/RespSortedSetTests.cs:1164]: Please add test case where destination key is not empty

**Review by TalZaccai — CHANGES_REQUESTED** [review]
Great work! Most comments are similar to those on the ZINTER PR.

File: libs/server/Storage/Session/ObjectStore/SortedSetOps.cs (hunk 2 above)
Comment (Vijay-Nirmal) [reply libs/server/Storage/Session/ObjectStore/SortedSetOps.cs:1016]: It's already in the right format only

**Comment by Vijay-Nirmal** [comment]
@TalZaccai Fixed review commands

File: libs/server/Resp/Objects/SortedSetCommands.cs [hunk 4]
```diff
@@ -879,5 +880,228 @@ private unsafe bool SortedSetDifferenceStore<TGarnetApi>(ref TGarnetApi storageA
...
+
+            // Number of keys
+            if (!parseState.TryGetInt(0, out var nKeys))
+            {
+                return AbortWithErrorMessage(CmdStrings.RESP_ERR_GENERIC_VALUE_IS_NOT_INTEGER);
+            }
+
+            if (nKeys < 1)
+            {
+                return AbortWithErrorMessage(Encoding.ASCII.GetBytes(string.Format(CmdStrings.GenericErrAtLeastOneKey, "zunion")));
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SortedSetCommands.cs:906]: nit: nameof(RespCommand.ZUNION)

File: libs/server/Resp/Objects/SortedSetCommands.cs [hunk 5]
```diff
@@ -879,5 +880,228 @@ private unsafe bool SortedSetDifferenceStore<TGarnetApi>(ref TGarnetApi storageA
...
+                        while (!RespWriteUtils.WriteEmptyArray(ref dcurr, dend))
+                            SendAndReset();
+                        break;
+                    }
+
+                    // write the size of the array reply
+                    while (!RespWriteUtils.WriteArrayLength(includeWithScores ? result.Count * 2 : result.Count, ref dcurr, dend))
+                        SendAndReset();
+
+                    if (result != null)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SortedSetCommands.cs:989]: no need for this condition now

File: libs/server/SessionParseStateExtensions.cs [hunk 6]
```diff
@@ -213,5 +213,21 @@ internal static bool TryGetExpireOption(this SessionParseState parseState, int i
 
             return true;
         }
+
+        internal static bool TryGetSortedSetAggregateType(this SessionParseState parseState, int idx, out SortedSetAggregateType value)
```
Comment (TalZaccai) [inline libs/server/SessionParseStateExtensions.cs:217]: Please add an XML comment

**Review by TalZaccai — COMMENTED** [review]
Thanks for addressing all the comments! Couple final nits & one small comment still open from previous review

File: libs/server/Resp/Objects/SortedSetCommands.cs (hunk 4 above)
Comment (Vijay-Nirmal) [reply libs/server/Resp/Objects/SortedSetCommands.cs:1054]: Done

File: libs/server/Resp/Objects/SortedSetCommands.cs (hunk 5 above)
Comment (Vijay-Nirmal) [reply libs/server/Resp/Objects/SortedSetCommands.cs:1137]: Done

File: libs/server/SessionParseStateExtensions.cs (hunk 6 above)
Comment (Vijay-Nirmal) [reply libs/server/SessionParseStateExtensions.cs:217]: Done

**Comment by Vijay-Nirmal** [comment]
Tested all review comments. @TalZaccai I couldn't find any "comment still open from previous review"

**Review by TalZaccai — APPROVED** [review]

**Review by TalZaccai — COMMENTED** [review]
Please resolve conflicts and I’ll merge this in @Vijay-Nirmal :)

**Comment by Vijay-Nirmal** [comment]
Fixed conflicts
